crates/ruma-events/src/room/message/url_preview.rs

```diff
@@ -0,0 +1,542 @@
+            kty: "oct".to_owned(),
+            key_ops: vec!["encrypt".to_owned(), "decrypt".to_owned()],
+            alg: "A256CTR".to_owned(),
+            k: Base64::new(vec![0; 64]),
+            ext: true,
+        }
+    }
+
+    fn encrypted_file() -> EncryptedFile {
+        let mut hashes: BTreeMap<String, Base64> = BTreeMap::new();
+        hashes.insert("sha256".to_owned(), Base64::new(vec![1; 10]));
+        EncryptedFile {
+            url: owned_mxc_uri!("mxc://localhost/encryptedfile"),
+            key: dummy_jwt(),
+            iv: Base64::new(vec![1; 12]),
+            hashes,
+            v: "v2".to_owned(),
+        }
+    }
+
+    #[test]
+    fn serialize_preview_image() {
+        let expected_result = json!({
+            "og:image": "mxc://maunium.net/zeHhTqqUtUSUTUDxQisPdwZO"
+        });
+
+        let preview =
+            PreviewImage::plain(owned_mxc_uri!("mxc://maunium.net/zeHhTqqUtUSUTUDxQisPdwZO"));
+
+        assert_eq!(to_json_value(&preview).unwrap(), expected_result);
+
+        let encrypted_result = json!({
+            "beeper:image:encryption": {
+                "hashes" : {
+                    "sha256": "AQEBAQEBAQEBAQ",
+                },
+                "iv": "AQEBAQEBAQEBAQEB",
+                "key": {
+                    "alg": "A256CTR",
+                    "ext": true,
+                    "k": "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
+                    "key_ops": [
+                        "encrypt",
+                        "decrypt"
+                    ],
+                    "kty": "oct",
+                },
+                "v": "v2",
+                "url": "mxc://localhost/encryptedfile",
+            },
+        });
+
+        let preview = PreviewImage::encrypted(encrypted_file());
+
+        assert_eq!(to_json_value(&preview).unwrap(), encrypted_result);
+    }
+
+    #[test]
+    fn serialize_room_message_with_url_preview() {
+        let expected_result = json!({
+            "msgtype": "m.text",
+            "body": "Test message",
+            "com.beeper.linkpreviews": [
+                {
+                    "matched_url": "https://matrix.org/",
+                    "og:image": "mxc://maunium.net/zeHhTqqUtUSUTUDxQisPdwZO",
+                }
+            ]
+        });
+
+        let preview_img =
+            PreviewImage::plain(owned_mxc_uri!("mxc://maunium.net/zeHhTqqUtUSUTUDxQisPdwZO"));
+        let full_preview = assign!(UrlPreview::matched_url("https://matrix.org/".to_owned()), {image: Some(preview_img)});
+        let msg = MessageType::Text(assign!(TextMessageEventContent::plain("Test message"),  {
+            url_previews: Some(vec![full_preview])
+        }));
+
+        assert_eq!(to_json_value(&RoomMessageEventContent::new(msg)).unwrap(), expected_result);
+    }
+
+    #[test]
+    fn serialize_room_message_with_url_preview_with_encrypted_image() {
+        let expected_result = json!(
```

Comment:
rust-analyzer insists on this indentation, if I switch this to `json!({` and remove one level of indentation, the next save my rust-analyzer indents it again to 8 ...

Reply:
That's because L234 has the wrong indentation. Since it's a long string anyway, it doesn't have to be on its own line, you can leave it on L233.